File: scripts/http-vsock-proxy.js
const http = require('http');
const net = require('net');
const fs = require('fs');

const PORT = 8000;
const ENCLAVE_CID = 16;
const ENCLAVE_PORT = 5000;
const AF_VSOCK = 40;
const VMADDR_CID_ANY = -1;

function createVsockConnection(cid, port, callback) {
  const { spawn } = require('child_process');

  const nc = spawn('nc', ['--vsock', cid.toString(), port.toString()]);

  nc.on('error', callback);
  nc.on('spawn', () => callback(null, nc));

  return nc;
}

const server = http.createServer((req, res) => {
  if (req.method !== 'POST') {
    res.writeHead(405);
    res.end('Method not allowed');
    return;
  }

  let body = '';
  req.on('data', chunk => body += chunk);
  req.on('end', () => {
    try {
      const message = JSON.parse(body);
      const messageStr = JSON.stringify(message) + '\n';

      const proc = spawn('nc', ['--vsock', ENCLAVE_CID.toString(), ENCLAVE_PORT.toString()]);

      proc.stdin.write(messageStr);
      proc.stdin.end();

      let response = '';
      proc.stdout.on('data', data => response += data.toString());

      proc.on('close', () => {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(response.trim() || '{}');
      });

      proc.on('error', err => {
        res.writeHead(500);
        res.end(JSON.stringify({ error: err.message }));
      });
    } catch (err) {
      res.writeHead(400);
      res.end(JSON.stringify({ error: err.message }));
    }
  });
});

server.listen(PORT, () => {
  console.log(`HTTP→vsock proxy on :${PORT} → vsock:${ENCLAVE_CID}:${ENCLAVE_PORT}`);
});

File: scripts/http-vsock-proxy.py
#!/usr/bin/env python3
import socket
import json
from http.server import HTTPServer, BaseHTTPRequestHandler

ENCLAVE_CID = 16
ENCLAVE_PORT = 5000
HTTP_PORT = 8000

class VsockProxy(BaseHTTPRequestHandler):
    def do_POST(self):
        content_length = int(self.headers['Content-Length'])
        body = self.rfile.read(content_length)

        try:
            message = json.loads(body)
            s = socket.socket(socket.AF_VSOCK, socket.SOCK_STREAM)
            s.connect((ENCLAVE_CID, ENCLAVE_PORT))
            s.sendall((json.dumps(message) + '\n').encode())

            response = b''
            while True:
                chunk = s.recv(4096)
                if not chunk:
                    break
                response += chunk
                if b'\n' in chunk:
                    break

            s.close()

            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(response.strip())

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())

    def log_message(self, format, *args):
        print(f"{self.address_string()} - {format % args}")

if __name__ == '__main__':
    server = HTTPServer(('0.0.0.0', HTTP_PORT), VsockProxy)
    print(f'HTTP→vsock proxy on :{HTTP_PORT} → vsock:{ENCLAVE_CID}:{ENCLAVE_PORT}')
    server.serve_forever()
